core/services/gateway/connectionmanager.go

```diff
@@ -72,40 +117,155 @@ func (m *connectionManager) DONConnectionManager(donId string) DONConnectionMana
-	// TODO (FUN-469): implement the handshake (extract node address, validate signature, create a new attempt, generate challenge)
-	return "", []byte{}, nil
+	nodeAddress, authHeaderElems, err := m.parseAuthHeader(authHeader)
+	if err != nil {
+		return "", nil, err
+	}
+	donConnMgr, ok := m.dons[authHeaderElems.DonId]
+	if !ok {
+		return "", nil, errors.New("invalid DON ID")
+	}
+	nodeState, ok := donConnMgr.nodes[nodeAddress]
+	if !ok {
+		return "", nil, errors.New("no such node")
+	}
+	nowTs := uint32(m.clock.Now().Unix())
+	ts := authHeaderElems.Timestamp
+	if ts < nowTs-m.config.AuthTimestampToleranceSec || nowTs+m.config.AuthTimestampToleranceSec < ts {
+		return "", nil, errors.New("timestamp out of tolerance zone")
+	}
+	nodeState.mu.RLock()
+	lastAcceptedTs := nodeState.lastAcceptedTs
+	nodeState.mu.RUnlock()
+	if ts <= lastAcceptedTs {
+		return "", nil, errors.New("timestamp too low")
+	}
+	attemptId, challenge, err = m.newAttempt(nodeState, nodeAddress, ts)
+	if err != nil {
+		return "", nil, err
+	}
+	return attemptId, challenge, nil
+}
+
+func (m *connectionManager) parseAuthHeader(authHeader []byte) (nodeAddress string, authHeaderElems *network.AuthHeaderElems, err error) {
+	n := len(authHeader)
+	if n < network.HandshakeAuthHeaderMinLen {
+		return "", nil, errors.New("auth header too short")
+	}
+	authHeaderElems, err = network.Unpack(authHeader[:n-network.HandshakeSignatureLen])
+	if err != nil {
+		return "", nil, errors.New("unable to parse auth header")
+	}
+	signature := authHeader[n-network.HandshakeSignatureLen:]
+	signer, err := ValidateSignature(signature, authHeader[:n-network.HandshakeSignatureLen])
+	nodeAddress = "0x" + hex.EncodeToString(signer)
+	return
+}
+
+func (m *connectionManager) newAttempt(nodeSt *nodeState, nodeAddress string, timestamp uint32) (string, []byte, error) {
+	challenge := make([]byte, m.config.AuthChallengeLen)
+	_, err := rand.Read(challenge)
+	if err != nil {
+		return "", nil, err
+	}
+	m.connAttemptsMu.Lock()
+	defer m.connAttemptsMu.Unlock()
+	m.connAttemptCounter++
+	newId := fmt.Sprintf("%s_%d", nodeAddress, m.connAttemptCounter)
+	m.connAttempts[newId] = &connAttempt{nodeState: nodeSt, nodeAddress: nodeAddress, challenge: challenge, timestamp: timestamp}
+	return newId, challenge, nil
 }
 
 func (m *connectionManager) FinalizeHandshake(attemptId string, response []byte, conn *websocket.Conn) error {
-	m.lggr.Debug("FinalizeHandshake", attemptId, response)
-	// TODO (FUN-469): implement the handshake (validate signature, add a new connection)
+	m.lggr.Debug("FinalizeHandshake attempt: ", attemptId)
+	m.connAttemptsMu.Lock()
+	attempt, ok := m.connAttempts[attemptId]
+	m.connAttemptsMu.Unlock()
+	if !ok {
+		return errors.New("connection attempt not found")
+	}
+	signer, err := ValidateSignature(response, attempt.challenge)
+	if err != nil {
+		return errors.New("invalid challenge response")
+	}
+	if attempt.nodeAddress != "0x"+hex.EncodeToString(signer) {
```

Comment:
A little concern here is that the case will match. Can nodeAddress be a checksum address? 
https://coincodex.com/article/2078/ethereum-address-checksum-explained/

I would convert to `common.Address` and compare bytes.

Reply:
attempt.nodeAddress is generated in parseAuthHeader out of bytes so it's always lowercase, just like the right-hand side.